sway-core/src/type_system/info.rs

```diff
@@ -1526,6 +1526,246 @@ impl TypeInfo {
     pub fn is_unknown_generic(&self) -> bool {
         matches!(self, TypeInfo::UnknownGeneric { .. })
     }
+
+    pub fn new_tuple(engines: &Engines, items: impl IntoIterator<Item = TypeInfo>) -> TypeInfo {
+        let te = engines.te();
+        let items = items
+            .into_iter()
+            .map(|x| te.insert(engines, x, None))
```

Comment:
Quite some time ago @JoshuaBatty, @tritao and I were discussing internally, that passing the `source_id` to the `insert` is error prone because it forces all callers to understand the semantics of it and to pass the proper `source_id` which is actually dependent of the type of the `TypeInfo` being inserted etc. We had already places in code where wrong `source_id` information was provided. I wanted to refactor the inserting logic and make it robust but never found time for that task :disappointed: 

My question is here, @JoshuaBatty and @tritao I hope you can provide some more insights, what will happen if `items` contain types that should have `source_id` like e.g. structs. Here we are setting `source_id` to `None` for all types. Can this cause any issues in the garbage collection or can it be ignored?

The `new_tuple` function is new and currently used only to create `(ptr, u64, u64)` where the `source_id` is `None` for all tuple elements. But since it is a general purpose function that can be used in future, I just want to be sure that we are not letting the door ajar for strange bugs down the road.

Reply:
Thanks @tritao! On the long term, the solution will very likely be refactoring `insert` to take care of those details plus special behavior for autogenerated code if needed. Right now, since `new_tuple` is used only once for types that do not have source, it looks to me that the easiest way to stay safe is to remove the function from `TypeInfo` and have it as a local private function in the file where it is actually used to create the buffer tuple `(ptr, u64, u64)`.

@xunilrj What do you think?

Reply:
Done.

Reply:
Thanks for checking in about the correct usage here @ironcev! Agree with @tritao's advice.